src/sentry/snuba/metrics/fields/base.py

```diff
@@ -325,6 +353,128 @@ def generate_default_null_values(self):
+                continue
+            constituent_metric_obj = DERIVED_METRICS[metric_name]
+            if isinstance(constituent_metric_obj, SingularEntityDerivedMetric):
+                entity = constituent_metric_obj.get_entity(projects=projects)
+                entities_and_metric_names.setdefault(entity, []).append(
+                    constituent_metric_obj.metric_name
+                )
+            entities_and_metric_names.update(
+                cls.__recursively_generate_singular_entity_constituents(
+                    projects, constituent_metric_obj
+                )
+            )
+
+        return entities_and_metric_names
+
+    def generate_bottom_up_derived_metrics_dependencies(self):
+        """
+        Function that builds a metrics dependency list from a derived metric_tree
+        As an example, let's consider the `session.errored` derived metric
+
+            session.errored
+               /   \
+              /    session.errored_preaggregated
+             /
+        session.errored_set
+
+        This function would generate a bottom up dependency list that would look something like
+        this ["session.errored_set", "session.errored_preaggregated", "session.errored"]
+
+        This is necessary especially for instances of `CompositeEntityDerivedMetric` because these
+        do not have a direct mapping to a query alias but are rather computed post query, and so to
+        calculate the value of that field we would need to guarantee that the values of its
+        constituent metrics are computed first.
+
+        This is more apparent when the dependency tree contains multiple levels of instances of
+        `CompositeEntityDerivedMetric`. Following up with our example,
+
+        session.errored
+               /   \
+              /    composite_entity_derived_metric
+             /       \
+            /        session.errored_preaggregated
+           /
+        session.errored_set
+
+        In this modified example, our dependency list would change to
+        [
+        "session.errored_set", "session.errored_preaggregated",
+        "composite_entity_derived_metric", "session.errored"
+        ]
+        And this order is necessary because we would loop over this list to compute
+        `composite_entity_derived_metric` first which does not have a direct mapping to a query
+        alias before we are able to compute `session.errored` which also does not have a direct
+        query alias
+        """
+        import queue
+
+        results = []
+        queue = queue.Queue()
+        queue.put(self)
+        while not queue.empty():
+            node = queue.get()
+            if node.metric_name in DERIVED_METRICS:
+                results.append((None, node.metric_name))
+            for metric in node.metrics:
+                if metric in DERIVED_METRICS:
+                    queue.put(DERIVED_METRICS[metric])
+        return list(reversed(results))
```

Comment:
Nice! I like the dynamic programming approach. Any reason this is not a recursive function like the others though? I'm not sure what the overhead of creating a `Queue` is here.

Reply:
Just out of convenience, shouldn't they have similar performance? O(n) for time and space complexity in both cases?

Reply:
Thanks for raising this up.. you are right about the queue adding some overhead because its also in the context of a thread synchronization primitive. also a list is not a good idea here because I am initially doing left side pops which is a O(n) on a list.
After some research, probably the correct data structure to use here is a deque which is a lightweight Queue